File: implementations/java/server/src/main/java/com/theshuai/specusserver/management/model/HttpBodyDataCodec.java
package com.theshuai.specusserver.management.model;

import org.brotli.dec.BrotliInputStream;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.Locale;
import java.util.zip.GZIPInputStream;
import java.util.zip.Inflater;
import java.util.zip.InflaterInputStream;

public final class HttpBodyDataCodec {
    private HttpBodyDataCodec() {
    }

    public static String toDisplayText(byte[] bodyData, String contentType, String headers, String fallbackText) {
        if (bodyData == null || bodyData.length == 0) {
            return fallbackText == null ? "" : fallbackText;
        }
        String contentEncoding = headerValue(headers, "content-encoding");
        if (hasEncodedBody(contentEncoding)) {
            byte[] decoded = decodeContentEncoding(bodyData, contentEncoding);
            if (decoded != null) {
                return toDisplayText(decoded, contentType);
            }
            return dataUrl("application/octet-stream", bodyData);
        }
        return toDisplayText(bodyData, contentType);
    }

    private static String toDisplayText(byte[] bodyData, String contentType) {
        if (!isTextBody(contentType) && !looksLikeText(bodyData)) {
            return dataUrl(mediaType(contentType), bodyData);
        }
        return sanitizeText(new String(bodyData, StandardCharsets.UTF_8));
    }

    private static byte[] decodeContentEncoding(byte[] bodyData, String contentEncoding) {
        String[] tokens = contentEncoding.split(",");
        byte[] current = bodyData;
        try {
            for (int i = tokens.length - 1; i >= 0; i--) {
                String token = tokens[i].trim().toLowerCase(Locale.ROOT);
                if (token.isBlank() || token.equals("identity")) {
                    continue;
                }
                if (token.equals("gzip") || token.equals("x-gzip")) {
                    current = gunzip(current);
                    continue;
                }
                if (token.equals("deflate") || token.equals("x-deflate")) {
                    current = inflate(current);
                    continue;
                }
                if (token.equals("br")) {
                    current = brotli(current);
                    continue;
                }
                return null;
            }
            return current;
        } catch (IOException | IllegalArgumentException ignored) {
            return null;
        }
    }

    private static byte[] gunzip(byte[] data) throws IOException {
        try (GZIPInputStream input = new GZIPInputStream(new ByteArrayInputStream(data))) {
            return readAll(input);
        }
    }

    private static byte[] inflate(byte[] data) throws IOException {
        try {
            try (InflaterInputStream input = new InflaterInputStream(new ByteArrayInputStream(data))) {
                return readAll(input);
            }
        } catch (IOException first) {
            try (InflaterInputStream input = new InflaterInputStream(new ByteArrayInputStream(data), new Inflater(true))) {
                return readAll(input);
            }
        }
    }

    private static byte[] brotli(byte[] data) throws IOException {
        try (BrotliInputStream input = new BrotliInputStream(new ByteArrayInputStream(data))) {
            return readAll(input);
        }
    }

    private static byte[] readAll(java.io.InputStream input) throws IOException {
        ByteArrayOutputStream output = new ByteArrayOutputStream();
        byte[] buffer = new byte[8192];
        int read;
        while ((read = input.read(buffer)) >= 0) {
            output.write(buffer, 0, read);
        }
        return output.toByteArray();
    }

    private static String dataUrl(String mediaType, byte[] bodyData) {
        return "data:" + mediaType + ";base64," + Base64.getEncoder().encodeToString(bodyData);
    }

    private static String headerValue(String headers, String name) {
        if (headers == null || headers.isBlank()) {
            return null;
        }
        String normalizedName = name.toLowerCase(Locale.ROOT);
        for (String header : headers.split("\\R")) {
            int separator = header.indexOf(':');
            if (separator > 0 && normalizedName.equals(header.substring(0, separator).trim().toLowerCase(Locale.ROOT))) {
                return header.substring(separator + 1).trim();
            }
        }
        return null;
    }

    private static boolean hasEncodedBody(String contentEncoding) {
        if (contentEncoding == null || contentEncoding.isBlank()) {
            return false;
        }
        for (String token : contentEncoding.split(",")) {
            String normalized = token.trim().toLowerCase(Locale.ROOT);
            if (!normalized.isBlank() && !normalized.equals("identity")) {
                return true;
            }
        }
        return false;
    }

    private static boolean isTextBody(String contentType) {
        String mediaType = mediaType(contentType);
        return mediaType.startsWith("text/")
                || mediaType.equals("application/json")
                || mediaType.endsWith("+json")
                || mediaType.equals("application/xml")
                || mediaType.endsWith("+xml")
                || mediaType.equals("application/x-www-form-urlencoded")
                || mediaType.equals("application/graphql")
                || mediaType.equals("application/javascript")
                || mediaType.equals("application/ecmascript")
                || mediaType.equals("application/x-yaml")
                || mediaType.equals("application/yaml");
    }

    private static boolean looksLikeText(byte[] data) {
        int inspected = Math.min(data.length, 512);
        int controls = 0;
        for (int i = 0; i < inspected; i++) {
            int value = data[i] & 0xff;
            if (value == 0) {
                return false;
            }
            if (value < 0x20 && value != '\r' && value != '\n' && value != '\t') {
                controls++;
            }
        }
        return inspected == 0 || controls * 10 <= inspected;
    }

    private static String mediaType(String contentType) {
        if (contentType == null || contentType.isBlank()) {
            return "application/octet-stream";
        }
        String mediaType = contentType.split(";", 2)[0].trim().toLowerCase(Locale.ROOT);
        return mediaType.matches("[a-z0-9!#$&^_.+-]+/[a-z0-9!#$&^_.+-]+") ? mediaType : "application/octet-stream";
    }

    private static String sanitizeText(String text) {
        if (text == null || text.isEmpty()) {
            return "";
        }
        StringBuilder result = new StringBuilder(text.length());
        for (int i = 0; i < text.length(); i++) {
            char ch = text.charAt(i);
            if ((Character.isISOControl(ch) || Character.isSurrogate(ch)) && ch != '\r' && ch != '\n' && ch != '\t') {
                result.append('.');
            } else {
                result.append(ch);
            }
        }
        return result.toString();
    }
}

File: implementations/csharp/server/src/Specus.Server/Management/HttpBodyDataCodec.cs
using System.IO.Compression;
using System.Text;
using System.Text.RegularExpressions;

namespace Specus.Server.Management;

/// <summary>
/// Reconstructs an HTTP body for the detail API from the stored wire bytes. The database and
/// Elasticsearch representations deliberately retain the original compressed/binary payload;
/// this codec only decodes it when an operator opens a detail record.
/// </summary>
internal static partial class HttpBodyDataCodec
{
    public static string ToDisplayText(byte[]? bodyData, string? contentType,
        string? headers, string? fallbackText)
    {
        if (bodyData is null || bodyData.Length == 0)
        {
            return fallbackText ?? string.Empty;
        }

        var contentEncoding = HeaderValue(headers, "content-encoding");
        if (HasEncodedBody(contentEncoding))
        {
            var decoded = DecodeContentEncoding(bodyData, contentEncoding!);
            if (decoded is not null)
            {
                return ToDisplayText(decoded, contentType);
            }

            return DataUrl("application/octet-stream", bodyData);
        }
        return ToDisplayText(bodyData, contentType);
    }

    private static string ToDisplayText(byte[] bodyData, string? contentType)
    {
        if (!IsTextBody(contentType) && !LooksLikeText(bodyData))
        {
            return DataUrl(MediaType(contentType), bodyData);
        }
        return SanitizeText(Encoding.UTF8.GetString(bodyData));
    }

    private static byte[]? DecodeContentEncoding(byte[] bodyData, string contentEncoding)
    {
        var current = bodyData;
        try
        {
            foreach (var token in contentEncoding
                         .Split(',', StringSplitOptions.TrimEntries | StringSplitOptions.RemoveEmptyEntries)
                         .Reverse())
            {
                if (token.Equals("identity", StringComparison.OrdinalIgnoreCase))
                {
                    continue;
                }
                current = DecodeOne(current, token) ?? throw new InvalidDataException();
            }
            return current;
        }
        catch (Exception ex) when (ex is InvalidDataException or IOException or ArgumentException)
        {
            return null;
        }
    }

    private static byte[]? DecodeOne(byte[] data, string token)
    {
        if (token.Equals("gzip", StringComparison.OrdinalIgnoreCase)
            || token.Equals("x-gzip", StringComparison.OrdinalIgnoreCase))
        {
            return ReadDecoded(data, source => new GZipStream(source, CompressionMode.Decompress));
        }
        if (token.Equals("deflate", StringComparison.OrdinalIgnoreCase)
            || token.Equals("x-deflate", StringComparison.OrdinalIgnoreCase))
        {
            try
            {
                return ReadDecoded(data, source => new ZLibStream(source, CompressionMode.Decompress));
            }
            catch (InvalidDataException)
            {
                return ReadDecoded(data, source => new DeflateStream(source, CompressionMode.Decompress));
            }
        }
        if (token.Equals("br", StringComparison.OrdinalIgnoreCase))
        {
            return ReadDecoded(data, source => new BrotliStream(source, CompressionMode.Decompress));
        }
        return null;
    }

    /// <summary>
    /// Reading a decompressor to the end lets a few kilobytes of crafted input cost gigabytes of
    /// memory, so both an absolute size and an expansion ratio are enforced.
    /// </summary>
    private static byte[] ReadDecoded(byte[] data, Func<Stream, Stream> decoderFactory)
    {
        using var source = new MemoryStream(data);
        using var decoder = decoderFactory(source);
        return Specus.Server.Http.DecompressionLimits.ReadAllBounded(decoder, data.Length);
    }

    private static string? HeaderValue(string? headers, string name)
    {
        if (string.IsNullOrWhiteSpace(headers))
        {
            return null;
        }
        foreach (var header in headers.Split(['\r', '\n'], StringSplitOptions.RemoveEmptyEntries))
        {
            var separator = header.IndexOf(':', StringComparison.Ordinal);
            if (separator > 0 && header[..separator].Trim().Equals(name, StringComparison.OrdinalIgnoreCase))
            {
                return header[(separator + 1)..].Trim();
            }
        }
        return null;
    }

    private static bool HasEncodedBody(string? contentEncoding) =>
        !string.IsNullOrWhiteSpace(contentEncoding)
        && contentEncoding.Split(',', StringSplitOptions.TrimEntries | StringSplitOptions.RemoveEmptyEntries)
            .Any(token => !token.Equals("identity", StringComparison.OrdinalIgnoreCase));

    private static bool IsTextBody(string? contentType)
    {
        var media = MediaType(contentType);
        return media.StartsWith("text/", StringComparison.Ordinal)
               || media == "application/json"
               || media.EndsWith("+json", StringComparison.Ordinal)
               || media == "application/xml"
               || media.EndsWith("+xml", StringComparison.Ordinal)
               || media == "application/x-www-form-urlencoded"
               || media == "application/graphql"
               || media == "application/javascript"
               || media == "application/ecmascript"
               || media == "application/x-yaml"
               || media == "application/yaml";
    }

    private static bool LooksLikeText(byte[] data)
    {
        var inspected = Math.Min(data.Length, 512);
        var controls = 0;
        for (var index = 0; index < inspected; index++)
        {
            var value = data[index];
            if (value == 0)
            {
                return false;
            }
            if (value < 0x20 && value is not (byte)'\r' and not (byte)'\n' and not (byte)'\t')
            {
                controls++;
            }
        }
        return inspected == 0 || controls * 10 <= inspected;
    }

    private static string MediaType(string? contentType)
    {
        if (string.IsNullOrWhiteSpace(contentType))
        {
            return "application/octet-stream";
        }
        var media = contentType.Split(';', 2)[0].Trim().ToLowerInvariant();
        return MediaTypePattern().IsMatch(media) ? media : "application/octet-stream";
    }

    private static string DataUrl(string mediaType, byte[] bodyData) =>
        $"data:{mediaType};base64,{Convert.ToBase64String(bodyData)}";

    private static string SanitizeText(string text)
    {
        if (text.Length == 0)
        {
            return string.Empty;
        }
        var result = new StringBuilder(text.Length);
        foreach (var character in text)
        {
            result.Append(char.IsControl(character) && character is not '\r' and not '\n' and not '\t'
                || char.IsSurrogate(character)
                    ? '.'
                    : character);
        }
        return result.ToString();
    }

    [GeneratedRegex("^[a-z0-9!#$&^_.+-]+/[a-z0-9!#$&^_.+-]+$", RegexOptions.CultureInvariant)]
    private static partial Regex MediaTypePattern();
}
